tools/depcheck/pkg/graph/graph.go

```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
 	return nil
 }
 
+// RemoveNode deletes the given node from the graph,
+// removing all of its outbound and inbound edges as well.
+func (g *MutableDirectedGraph) RemoveNode(n *Node) {
+	delete(g.nodesByName, n.UniqueName)
+	g.DirectedGraph.RemoveNode(n)
+}
+
 func (g *MutableDirectedGraph) NodeByName(name string) (graph.Node, bool) {
 	n, exists := g.nodesByName[name]
 	return n, exists && g.DirectedGraph.Has(n)
 }
+
+// PruneOrphans recursively removes nodes with no inbound edges.
+// Receives a list of UniqueName prefixes of nodes to ignore.
+// Returns a list of recursively pruned nodes.
+func (g *MutableDirectedGraph) PruneOrphans(ignore []string) []*Node {
+	removed := []*Node{}
+
+	for _, n := range g.Nodes() {
+		node, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+		if len(g.To(n)) > 0 {
+			continue
+		}
+
+		skip := false
+		for _, toIgnore := range ignore {
```

Comment:
make a private, `contains(needle string, haystack []string)` so this doesn't look as ugly.